### PR TITLE
fix: improve tool type, bump pydantic and outlines

### PR DESCRIPTION
**Pull request by drbh**
This PR resolves a couple 

- [X] adjusts the tool response to align with openai's tools response type
- [X] bumps pydantic to `2.6.4` in all apps (resolves dependency issue when running tests)
- [X] bump `outlines` version and fix import for new name

### CODE REVIEW

**Comment by drbh** [comment]
this PR resolves
https://github.com/huggingface/text-generation-inference/issues/1624
https://github.com/huggingface/text-generation-inference/pull/1653

File: integration-tests/pyproject.toml [hunk 1]
```diff
@@ -5,6 +5,7 @@ description = "Text Generation Inference integration tests"
 authors = ["Nicolas Patry <nicolas@huggingface.co>"]
 
 [tool.poetry.dependencies]
+pydantic = "> 1.10, < 3"
```
Comment (Narsil) [inline integration-tests/pyproject.toml:8]:
```suggestion
pydantic = "> 2, < 3"
```
no ? 2 is not compatible with 1 iirc.

File: router/src/lib.rs [hunk 2]
```diff
@@ -781,7 +781,7 @@ pub(crate) struct Message {
     #[schema(example = "\"David\"")]
     pub name: Option<String>,
     #[serde(default, skip_serializing_if = "Option::is_none")]
-    pub tool_calls: Option<ToolCall>,
+    pub tool_calls: Option<Vec<ChatCompletionMessageToolCall>>,
```
Comment (Narsil) [inline router/src/lib.rs:784]: I kind of prefer `ToolCall` in readability (chatCompletionMessage make it look like java imho).

File: server/pyproject.toml [hunk 3]
```diff
@@ -34,7 +34,7 @@ peft = { version = "^0.9.0", optional = true }
 torch = { version = "^2.1.1", optional = true }
 scipy = "^1.11.1"
 pillow = "^10.0.0"
-outlines= { version = "^0.0.27", optional = true }
+outlines= { version = "0.0.36", optional = true }
```
Comment (Narsil) [inline server/pyproject.toml:37]: Let´ s keep `^`  to get minor for free.

**Review by Narsil — DISMISSED** [review]
Some nits.

File: integration-tests/models/test_tools_llama.py [hunk 4]
```diff
@@ -119,19 +119,21 @@ async def test_flash_llama_grammar_tools(flash_llama_grammar_tools, response_sna
...
-            "name": "tools",
-            "parameters": {
-                "format": "celsius",
-                "location": "New York, NY",
-                "num_days": 14,
+    assert response.choices[0].message.tool_calls == [
+        {
+            "function": {
+                "description": None,
+                "name": "tools",
```
Comment (joumenharzli) [inline integration-tests/models/test_tools_llama.py:126]: the "name" here should be get_current_weather

File: integration-tests/models/test_tools_llama.py [hunk 5]
```diff
@@ -159,19 +161,21 @@ async def test_flash_llama_grammar_tools_auto(
...
```
Comment (joumenharzli) [inline integration-tests/models/test_tools_llama.py:168]: also here the "name" here should be "get_current_weather"

**Comment by joumenharzli** [comment]
Thanks @drbh for your PR, I've noticed that you changed the type of "tool_calls" from an object to a list, which is great. 
Additionally, there's another issue where the returned function name is consistently "tools". 
is this issue will be addressed in this PR?

**Comment by drbh** [comment]
Hi @joumenharzli thanks for highlighting that issue! Unfortunately the current expected value is `"name": "tools"` because of the issue you noted above (the LLM is not specifying the function it chose in it's response). 

Changing the format of tool format is interesting, and in general the currently implementation is a bit rigid. That change is a bit outside of the scope of this PR and I'll open a new issue to continue this conversation: https://github.com/huggingface/text-generation-inference/issues/1657

**Pull request merged**